Support recovery

DDofDD opened the pull request:
fix: 修复preStop快照保存超时时产生不完整的dev快照，导致工作站启动失败；
feat: 新增工作站&线上服务 ComfyUI进程崩溃（OOM等原因导致）时自动重启能力。

buptwzj666 committed:
feat(preStop): 添加超时控制和强制终止机制

- 使用 multiprocessing 替代 threading，支持 SIGKILL 强制终止 save 进程
- 主进程预生成 snapshot_name，便于超时时清理未完成的快照
- 添加 cleanup_incomplete_save 方法清理临时文件和未完成目录
- 新增 PRESTOP_TIMEOUT 配置项（默认 280 秒）
- 优化代码结构：早返回模式、移除冗余的 _saving_snapshot_name

buptwzj666 committed:
feat: 添加健康检查增强功能

- entrypoint.bash: 使用 exec 让 Python 替换 bash 成为 init 进程
- process_manager.py: 添加 dmesg 日志记录和进程退出逻辑，优化日志输出
- backend_process_manager.py: 优化代码结构，细化异常处理，调整健康检查策略

buptwzj666 committed:
feat: 改进 Serverless API 崩溃检测和错误处理

- 添加 WebSocket on_error 和 on_close 回调，监控连接状态
- 等待 WebSocket client_id 时添加 30 秒超时
- 检测工作流执行期间 ComfyUI 崩溃，提供清晰的错误信息
- PRESTOP_TIMEOUT 默认值从 280 秒增加到 580 秒
- 仅在 WebSocket 关闭状态码异常时记录日志

buptwzj666 committed:
refactor: 简化 WebSocket 崩溃检测逻辑

- 删除冗余的 ws_closed 检查代码（实际由 on_ws_error 处理）
- 保留 on_ws_error 和 on_ws_close 回调用于等待 client_id 阶段的保护

CiCi503 commented on src/code/agent/constants.py:
```diff
@@ -101,6 +101,9 @@
 DEFAULT_INSTALL_TIMEOUT = 600  # 10min default timeout for custom nodes installation
 
+# PreStop 相关配置
+PRESTOP_TIMEOUT = int(os.getenv("PRESTOP_TIMEOUT", "280"))  # preStop 超时时间，默认 280 秒（留 20 秒缓冲）
```
PrestopTimeoutLimit 300s=>600s，280 可以调大一些了

CiCi503 commented on src/code/agent/services/process/comfyui_process_manager.py:
```diff
@@ -0,0 +1,147 @@
+        try:
+            response = requests.get(
+                f'http://127.0.0.1:{constants.BACKEND_PROCESS_PORT}',
+                timeout=self.HEALTH_CHECK_TIMEOUT
```
判断 ComfyUI alive，不 alive 就重启，要再谨慎一点。
之前遇到过推理过程中，8188 也 curl 不通。
建议：
- 调大 HEALTH_CHECK_TIMEOUT => 10s
- 连续 3 次 HEALTH_CHECK 失败，再判定 died
<img width="2774" height="1036" alt="image" src="https://github.com/user-attachments/assets/4b3ff247-1c9f-43b2-86b2-140c02ea0c62" />

CiCi503 replied:
是否能有一种更可靠的方式判断 comfyui 挂没挂呢？看进程是否存在是一种可选方案吗？
一旦我们误判了，还是会带来不小的负面影响的
比如：
- 工作空间正在执行，就变成重启中了
- 线上环境正在执行，请求就阻塞了

DDofDD replied:
1. 超时时间变10s
2. 提供两种判断is_alive机制：
第一种：进程不再存在（例如被OOM Killed了），则直接重启；
第二种：进程存在但是GET http://127.0.0.1:{constants.BACKEND_PROCESS_PORT}连续三次失败，认为服务不可用，则重启

CiCi503 commented on src/code/agent/services/serverlessapi/serverless_api_service.py:
```diff
@@ -572,6 +572,38 @@ def get_status_from_store(self, task_id: str):
+        Raises:
+            Exception: 等待超时时抛出异常
+        """
+        from services.management_service import ManagementService, BackendStatus
```
L591~L593 都放到 L595 之后吧，判定为 Rebooting 再有 import

DDofDD replied:
嗯？BackendStatus.REBOOTING常量和ManagementService()这个单例服务也得先import才能判断

CiCi503 commented on src/code/agent/test/unit/services/comfyui_process_manager_test.py:
再补一些 UT 吧：
- serverless_api 在 Rebooting 状态下的行为
- reboot 成功或失败时的状态机变化
- reboot 后，comfyui 进程成功拉起

DDofDD replied:
comfyui_process_manager_test是我mock的一个假的http server来模拟的comfyui进程管控，我已经加了比较详细的UT；

serverless_api相关，由于当前UT就是有问题的，所以期望后面加UT时统一系统地加；
管控状态机相关，由于当前整体状态机被Reboot破坏，缺少包含Reboot状态扩展的UT，所以也期望后面加UT时统一系统地加。

buptwzj666 committed:
test: add comprehensive unit tests for ComfyUIProcessManager

- Add 9 new test cases covering health check, process state, and special modes
- Create mock processes for unhealthy, slow response, and crash scenarios
- Improve test documentation with detailed run guide and timing info
- Fix test isolation by avoiding background health check thread interference
- All 13 tests passing (original 4 + new 9)

Test coverage improvements:
- HTTP health check: consecutive failures, recovery, timeout
- Process state: crash detection, readiness check, running state
- Special modes: CPU mode and REBOOTING status

New mock processes:
- mock_unhealthy_process.py: HTTP 500 for health check failure testing
- mock_slow_response_process.py: 15s delay for timeout testing
- mock_crash_process.py: exit code 137 for OOM simulation
